### PR TITLE
Spm v2

### PR DESCRIPTION
**Pull request by daniel-ortiz**
This is the revised version with only the essential code based on the numatop core. The sources were located on the vendor/spm subdirectory.

There are still some small details to fix, but the core functionality is implemented

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -19,6 +19,7 @@ install:
     - travis_retry sudo aptitude -f install -y qt5-default qttools5-dev-tools
     - travis_retry sudo aptitude -f install -y gcc-4.8 g++-4.8
     - travis_retry sudo apt-get -f install -y linux-libc-dev
+    - travis_retry sudo apt-get -f install -y libnuma-dev
```
Comment (jbreitbart) [inline .travis.yml]: Can you please add that dependency to the readme file as well?

File: CMakeLists.txt [hunk 2]
```diff
@@ -18,6 +18,9 @@ string( TOLOWER "${CMAKE_BUILD_TYPE}" CMAKE_BUILD_TYPE )
 
 ########
 # Compiler/linker options based on the different compilers
+ADD_DEFINITIONS(-std=c++11 )
```
Comment (jbreitbart) [inline CMakeLists.txt]: Do you really need this here? -std=c++11 is added to CXX_FLAGS just a few lines below. If you really need this: which compiler/cmake are you using?

File: CMakeLists.txt [hunk 3]
```diff
@@ -161,10 +172,18 @@ include_directories(vendor/spdlog/include/spdlog)
...
+# NUMA migration tool
+include_directories(vendor/spm/)
+
+# Boost
+include_directories(${BoostSourceDir})
+
 # fastlib
 include_directories(${libfast_path}/include/fast-lib)
 include_directories(${libfast_path}/include/fast-lib/external)
-########
```
Comment (jbreitbart) [inline CMakeLists.txt]: ? Could you add this again? It is supposed to end the includes section :)

File: include/AutopinPlus/Monitor/PageMigrate/Main.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+#include <AutopinPlus/PerformanceMonitor.h>
+#include <map>
+
+namespace AutopinPlus {
+namespace Monitor {
+namespace PageMigrate {
+
+/*!
+ * \brief Random performance counter
```
Comment (jbreitbart) [inline include/AutopinPlus/Monitor/PageMigrate/Main.h]: Random? I guess you forgot to update the comments in this file. Could you please take a closer look at all the files in this

File: src/AutopinPlus/Monitor/PageMigrate/Main.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+ * Copyright (C) 2015 Technische Universität München - LRR
+ *
+ * This file is licensed under the GNU General Public License Version 3
+ */
+
+#include <AutopinPlus/Monitor/PageMigrate/Main.h>
+
+#include <QTime>
+#include <stdlib.h>
+extern "C" {
```
Comment (jbreitbart) [inline src/AutopinPlus/Monitor/PageMigrate/Main.cpp]: Could you add

#ifdef __cplusplus
extern "C" {
#endif
....
#ifdef __cplusplus
}
#endif

to the header file and remove the extern "C" here?

File: src/AutopinPlus/Monitor/PageMigrate/Main.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+
+namespace AutopinPlus {
+namespace Monitor {
+namespace PageMigrate {
+
+Main::Main(QString name, const Configuration &config, AutopinContext &context)
+	: PerformanceMonitor(name, config, context) {
+	type = "random";
```
Comment (jbreitbart) [inline src/AutopinPlus/Monitor/PageMigrate/Main.cpp]: Again, please check for any c&p error in this file :)

File: src/AutopinPlus/Monitor/PageMigrate/Main.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	//Must be left, this value is read by the strategy
+	valtype = MAX;
+	// Set standard values
+
+
+	if (config.configOptionExists(name + ".sampling_period") == 1) period = config.getConfigOptionInt(name + ".sampling_period");
+	if (config.configOptionExists(name + ".min_weight") == 1) min_weight = config.getConfigOptionInt(name + ".min_weight");
+	if (config.configOptionExists(name + ".sensing_time") == 1) sensing_time = config.getConfigOptionInt(name + ".sensing_time");
+	
+	//context.info("Maximum random value " + QString::number(rand_max));
```
Comment (jbreitbart) [inline src/AutopinPlus/Monitor/PageMigrate/Main.cpp]: Just remove this line if you don't need it anymore.

File: src/AutopinPlus/Monitor/PageMigrate/Main.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	st.ll_sampling_period=period;
+	st.ll_weight_threshold=min_weight;
+	st.measure_time=sensing_time;
+	st.pid_uo=monitored_pid;
+	init_spm(&st);
+}
+
+double Main::value(int tid) {
+	
+	return tid;
```
Comment (jbreitbart) [inline src/AutopinPlus/Monitor/PageMigrate/Main.cpp]: Can you return any current performance data here and not just the tid?

File: src/AutopinPlus/Monitor/PageMigrate/Main.cpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+double Main::stop(int tid) {
+	double result = value(tid);
+	if (context.isError()) {
+		return 0;
+	}
+
+	rands.erase(tid);
```
Comment (jbreitbart) [inline src/AutopinPlus/Monitor/PageMigrate/Main.cpp]: What is this actually doing?

**Comment by jbreitbart** [comment]
Feel free to add yourself to the AUTHORS file if you want.

**Comment by jbreitbart** [comment]
I did not check the code in the vendor directory in details. Do you want me to take a look at anything specific?

Is it ok if we try the code code?

**Comment by jbreitbart** [comment]
@daniel-ortiz Is that the code you used for your measurements?

**Comment by daniel-ortiz** [comment]
Hello Jens

I have just commited the code I used for the measurements  integrated with autopin (vendor/spm folder).

But to be exact, the code I used was a standalone version not integrated with autopin (https://github.com/daniel-ortiz/spm-vII.git). They are both the same, the difference is the integration

**Comment by daniel-ortiz** [comment]
I updated the push request with the final code and all the suggestion you made to my last commit.

The only missing is a better documentation of the code, but besides that the functional part is finished.

**Pull request merged**